TESTS/netsocket/tcp/main.cpp

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2018, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_CONF_APP_CONNECT_STATEMENT
+#error [NOT_SUPPORTED] No network configuration found for this target.
+#endif
+
+#include "mbed.h"
+#include MBED_CONF_APP_HEADER_FILE
+#include "greentea-client/test_env.h"
+#include "unity/unity.h"
+#include "utest.h"
+#include "utest/utest_stack_trace.h"
+#include "tcp_tests.h"
+
+using namespace utest::v1;
+
+namespace
+{
+    NetworkInterface* net;
+}
+
+NetworkInterface* get_interface()
+{
+    return net;
+}
+
+void drop_bad_packets(TCPSocket& sock, int orig_timeout) {
+    nsapi_error_t err;
+    sock.set_timeout(0);
+    while (true) {
+        err = sock.recv(NULL, 0);
+        if (err == NSAPI_ERROR_WOULD_BLOCK) {
+            break;
+        }
+    }
+    sock.set_timeout(orig_timeout);
+}
+
+static void _ifup() {
+    net = MBED_CONF_APP_OBJECT_CONSTRUCTION;
+    nsapi_error_t err = MBED_CONF_APP_CONNECT_STATEMENT;
+    TEST_ASSERT_EQUAL(NSAPI_ERROR_OK, err);
+    printf("MBED: TCPClient IP address is '%s'\n", net->get_ip_address());
+}
+
+static void _ifdown() {
+    net->disconnect();
+    printf("MBED: ifdown\n");
+}
+
+void tcpsocket_connect_to_echo_srv(TCPSocket& sock) {
+    SocketAddress tcp_addr;
+
+    get_interface()->gethostbyname(MBED_CONF_APP_ECHO_SERVER_ADDR, &tcp_addr);
```

Comment:
Hi
In previous tests, there was an option to not explicitly define server address and port,
and use greentea to get dynamically the used test PC (greentea_send_kv("host_ip",...).
Maybe it can be implemented ?
Thx

Reply:
The idea here is that we test against publicly available echo server, not one running on local PC.

We are no-more expecting test machine to be connected to same network as it is a limitation that cannot be done for some technologies, like Cellular for example.

Reply:
I agree, I don't want to modify the default public test.

Just want to add something like :

#if defined(MBED_CONF_APP_ECHO_SERVER_ADDR) && defined(MBED_CONF_APP_ECHO_SERVER_PORT)		
get_interface()->gethostbyname(MBED_CONF_APP_ECHO_SERVER_ADDR, &tcp_addr);
#else /* MBED_CONF_APP_ECHO_SERVER_ADDR && MBED_CONF_APP_ECHO_SERVER_PORT */		
/* use greentea_send_kv  and greentea_parse_kv */
get_interface()->gethostbyname(LOCAL_HOST, &tcp_addr);
#endif

Reply:
If you have local echo server, you might as well provide its address in `mbed_app.json` by overriding MBED_CONF_APP_ECHO_SERVER. No need for #ifdefs.

Reply:
Does anyone know how to set up a local TCP/UDP echo server ?
Thx

Reply:
Answered by email,

However this is very valid question.
We are preparing to release not only test environment documentation, but test case documentation as well.
There are so many expectation about environment and implementations that those cannot be guessed right by just looking at the code.